src/cpp/session/modules/SessionPackages.R

```diff
@@ -430,44 +433,93 @@ if (identical(as.character(Sys.info()["sysname"]), "Darwin") &&
+{
+   # first, check if we've already discovered a NEWS link
+   cache <- .rs.packageNewsURLsEnv
+   entry <- file.path(libraryPath, packageName)
+   if (exists(entry, envir = cache))
+      return(get(entry, envir = cache))
+   
+   # determine an appropriate CRAN URL
+   repos <- getOption("repos")
+   cran <- if ("CRAN" %in% names(repos))
+      repos[["CRAN"]]
+   else if (length(repos))
+      repos[[1]]
+   else
+      "https://cran.rstudio.com"
+   
+   # check to see if this package was from Bioconductor. if so, we'll need
+   # to construct a more appropriate url
+   desc <- .rs.tryCatch(.rs.readPackageDescription(file.path(libraryPath, packageName)))
+   prefix <- if (inherits(desc, "error") || !"biocViews" %in% names(desc))
+      file.path(cran, "web/packages")
    else
-      cranRep <- c(CRAN = repos[[1]])
-   cranRepLen <- nchar(cranRep)
-   isFromCRAN <- cranRep == substr(updates$Repository, 1, cranRepLen)
-   hasNEWS    <- file.exists(vapply(updates$Package, function(x) system.file("NEWS", package = x), character(1L)))
-   newsURL <- character(nrow(updates))
-   if (substr(cranRep, cranRepLen, cranRepLen) != "/")
-      cranRep <- paste(cranRep, "/", sep="")
-
-   newsURL[isFromCRAN] <- paste(cranRep,
-                                "web/packages/",
-                                updates$Package,
-                                ifelse(hasNEWS, "/NEWS", "/news.html"),
-                                sep = "")[isFromCRAN]
-   
-   updates <- data.frame(packageName = updates$Package,
-                         libPath = updates$LibPath,
-                         installed = updates$Installed,
-                         available = updates$ReposVer,
-                         newsUrl = newsURL,
-                         stringsAsFactors = FALSE)
-                       
-                       
-   return (updates)
+      "https://bioconductor.org/packages/release/bioc/news"
+   
+   # the set of candidate URLs -- we use the presence of a NEWS or NEWS.md
+   # to help us prioritize the order of checking.
+   #
+   # in theory, the current-installed package might not have NEWS at all, but
+   # the latest released version might have it after all, so checking the
+   # current installed package is just a heuristic and won't be accurate
+   # 100% of the time
+   pkgPath <- file.path(libraryPath, packageName)
+   candidates <- if (file.exists(file.path(pkgPath, "NEWS.md"))) {
+      c("news/news.html", "news.html", "NEWS", "ChangeLog")
+   } else if (file.exists(file.path(pkgPath, "NEWS"))) {
+      c("NEWS", "news/news.html", "news.html", "ChangeLog")
+   } else {
+      c("news/news.html", "news.html", "NEWS", "ChangeLog")
+   }
+   
+   for (candidate in candidates) {
+      
+      url <- file.path(prefix, packageName, candidate)
+      
+      status <- .rs.withTimeLimit(5, {
+         .rs.tryCatch(download.file(url, destfile = tempfile(), quiet = TRUE, mode = "wb"))
```

Comment:
I think we can use `HEAD` to test for the URL's existence without downloading the whole thing, although possibly not worth doing if R doesn't make it easy.

Reply:
Given that the NEWS files tend to be only a few kilobytes in size I think it should be okay to just attempt to download the file directly. (You're correct that there isn't a simple, dependency-less way of making a `HEAD` request from R)

I thought about whether it'd be worth exposing some of our C++ machinery for doing this, but using `download.file()` also ensures we respect all of the other download options set in the R session (`download.file.method`, `download.file.extra`, any options or environment variables affecting `curl` or `wget`, and so on)